### PR TITLE
Add article-analysis partial failure handling and structured POST errors

### LINKED ISSUE #217: [hyper-pm] MP-ART-ANALYSIS-007: Handle partial failure and idempotency with structured errors per chunk or article
**Ticket ID:** `MP-ART-ANALYSIS-007`  
**Priority:** P0  

---

# Handle partial failure and idempotency with structured errors per chunk or article

**Group:** Article Analysis Agent (`article-analysis-agent`)

## Summary

Ensures one article or one POST chunk failing does not unnecessarily fail the entire run when partial success mode applies, consistent with data-collection hardening philosophy. Clarifies idempotency: default runs skip already analyzed sources via GET `unanalyzed=true` when `reanalyze` is false, reanalyze runs with bounds upsert mentions and relevance per API behavior without duplicating rows.

## Scope

- In scope: Per-article or per-chunk error isolation configuration. On LLM failure, mark affected articles failed with diagnostic context and continue when partial mode is enabled. On POST partial failure, surface API errors, do not claim success for counts the API did not confirm (FR7, Error Handling). Retry policy for transient errors if consistent with other agents. Document merge behavior for reanalyze with MP-ART-ANALYSIS-001 handler.
- Out of scope: New operator policy UI for discard versus rollback of prior chunks beyond logging and metrics.

## Acceptance criteria

- [x] Partial failure behavior matches documented mode (for example continue after recoverable extraction failure when enabled).
- [x] Successful chunks before a failed chunk are not reported as failed in aggregate metrics.
- [ ] Re-run with `reanalyze=false` does not reprocess already analyzed rows per GET semantics.
- [ ] Re-run with `reanalyze=true` and bounds overwrites or upserts per API without duplicate primary rows for the same logical keys.
- [ ] Structured logs or errors identify failing `dataSourceId` or chunk index without leaking secrets or full article body.
- [ ] Unit or integration tests cover partial failure and idempotent GET skip paths.

## Dependencies

- Depends on: MP-ART-ANALYSIS-006
- Blocks: MP-ART-ANALYSIS-008

## Notes

Coordinate exact upsert semantics with MP-ART-ANALYSIS-001 Prisma implementation. Align with PRD POST partial failure wording: surface API error and avoid false success claims.

PRD: https://github.com/hyperjumptech/mediapulse/blob/prds/article-analysis-agent-prd.md

```json
{
  "hyper_pm_id": "01KP1FZ60DQAZPJY9FHWZXA9QD",
  "type": "ticket",
  "parent_ids": {
    "epic": "01KP1JW1N0W9M73YC82WEWP5YY",
    "story": "01KP1JW2ZWMRP3G4YMP00KKQT4"
  },
  "depends_on": [
    "01KP1FZ60DWR5GBSMXGGCCRJ7G"
  ]
}
```

### PR DESCRIPTION
**Pull request by kevinhermawan**
## Summary

Adds MP-ART-ANALYSIS-007: configurable **`runPolicy`** (minimum successful extractions before POST), per-source **`extractionFailures`**, per-chunk **`postFailures`** with safe abort of later POST phases, optional **429/5xx retries** on `analysis.create`, and **`details.runStatus`** (`partial_success` vs success). Documents idempotency and fixes related tests (agent-data-api-client GET mocks, `loadAnalysisContext` stubs with `articleRelevance.count`).

## Related issues

Refs #217

## Important changes

- **article-analysis `run.ts`:** vocabulary/LLM skips recorded; POST try/catch + phase gating; honest chunk/count metrics.
- **Hermes config:** `runPolicy`, `postTransientRetries`, `postTransientRetryBaseDelayMs`.

## Other changes

- **agent-data-api:** idempotent `articleRelevance` upsert test; analysis route test mock includes `relevanceSelectionState`.
- **agent-data-api-client:** analysis GET test body includes `relevanceSelectionState`.
- **dev-docs:** `analysis.mdx` partial failure and idempotency.

## Key files to review

- `apps/mediapulse/agents/article-analysis/src/run.ts` — orchestration and details shape.
- `apps/mediapulse/agents/article-analysis/src/article-analysis-run-policy.ts` — policy + run status label.
- `apps/mediapulse/agents/article-analysis/src/article-analysis-agent-data-api-post.ts` — error parse + retries.
- `apps/mediapulse/agents/article-analysis/src/config-schema.ts` — new config fields.

## How to test

1. From repo root: `pnpm code-quality`.
2. Run article-analysis with Hermes `runPolicy` / partial extraction and confirm `details.extractionFailures` and `postFailures`; induce a failed ER chunk and assert mentions/relevance are not posted.

### CODE REVIEW

File: apps/mediapulse/agents/article-analysis/src/run.ts [hunk 1]
```diff
@@ -222,14 +248,44 @@ export const run = async ({
...
       } catch (err) {
-        llmFailures += 1;
+        const message = err instanceof Error ? err.message : String(err);
+        extractionFailures.push({
+          dataSourceId: source.id,
+          stage: "llm",
+          message,
+        });
         logger.warn(
           { err, dataSourceId: source.id, tickerId: input.tickerId },
```
Comment (nicnocquee) [inline apps/mediapulse/agents/article-analysis/src/run.ts]: We log the full `err` object here. For #217 we promised diagnostics without dragging article text into logs—some LLM clients stuff snippets into `cause` / `response`. I would mirror the vocabulary branch: structured fields only (`dataSourceId`, `stage: "llm"`, `message` from the safe string we already put in `extractionFailures`).

File: apps/mediapulse/agent-data-api/src/services/analysis.test.ts [hunk 2]
```diff
@@ -350,4 +359,57 @@ describe("applyAnalysisPost", () => {
...
+          },
+          selected: false,
+        },
+      ],
+    };
+
+    await applyAnalysisPost(body, { db: db as never });
+    await applyAnalysisPost(body, { db: db as never });
+
+    expect(tx.articleRelevance.upsert).toHaveBeenCalledTimes(2);
```
Comment (nicnocquee) [inline apps/mediapulse/agent-data-api/src/services/analysis.test.ts:413]: This proves we call `upsert` twice, which is useful, but it does not pin the acceptance line about no duplicate primary rows for the same logical keys. I would either assert the `where` / `create` / `update` shapes stay on the composite key we care about, or add a small DB-backed test if we want real duplicate insurance.

File: apps/mediapulse/agents/article-analysis/src/run.test.ts [hunk 3]
```diff
@@ -305,6 +313,189 @@ describe("run", () => {
     });
   });
 
+  it("continues after vocabulary failure on one source when another succeeds (partial_success)", async () => {
```
Comment (nicnocquee) [inline apps/mediapulse/agents/article-analysis/src/run.test.ts:639]: The PR docs say GET uses `unanalyzed=true` so incremental runs skip already-scored articles. There is no test for that.

Recommendation: add something like this in `run.test.ts`:

```ts
it("returns early when GET returns no data sources (everything already analyzed)", async () => {
  // mock analysis.get to return dataSources: [] — simulates everything already analyzed
});
```

(Flesh out assertions: e.g. no `analysis.create`, expected `success` / message / details.)

**Review by nicnocquee — CHANGES_REQUESTED** [review]

File: apps/mediapulse/agents/article-analysis/src/run.ts (hunk 1 above)
Comment (kevinhermawan) [reply apps/mediapulse/agents/article-analysis/src/run.ts]: Implemented in `run.ts`: LLM extraction failure logging now emits structured safe fields only (`tickerId`, `dataSourceId`, `stage: "llm"`, and sanitized `message`) and no longer logs the raw `err` object.

File: apps/mediapulse/agent-data-api/src/services/analysis.test.ts (hunk 2 above)
Comment (kevinhermawan) [reply apps/mediapulse/agent-data-api/src/services/analysis.test.ts:413]: Implemented in `analysis.test.ts`: the idempotency test now asserts the exact composite key shape (`where.dataSourceId_tickerId`) and verifies create/update payload shape on `articleRelevance.upsert`, so duplicate logical keys are explicitly pinned by test expectations.

File: apps/mediapulse/agents/article-analysis/src/run.test.ts (hunk 3 above)
Comment (kevinhermawan) [reply apps/mediapulse/agents/article-analysis/src/run.test.ts:639]: Added explicit run coverage in `run.test.ts` for incremental no-op behavior: `returns early when incremental GET has no data sources`. It asserts `analysis.get` receives `unanalyzed: true`, run succeeds with no-op message, and no POST call is made.

**Review by nicnocquee — APPROVED** [review]
